### DISCUSSION TITLE
Schrijfwijze van foutmeldingen

### DISCUSSION THREAD

**Discussion by jeffreylauwers** [nl-design-system/backlog#363] [Richtlijnen]
# Schrijfwijze van foutmeldingen

## Doel
Het zou goed zijn als we tot patronen kunnen komen voor het schrijven van foutmeldingen. In deze Discussie kunnen we daar aan samenwerken.

## Gerelateerde componenten

- [Form Field Error Message](https://github.com/orgs/nl-design-system/discussions/228)
- [Invalid Form Alert](https://github.com/orgs/nl-design-system/discussions/278)

## Richtlijnen

[Foutmeldingen in een formulier - NL Design System](https://www.nldesignsystem.nl/richtlijnen/formulieren/foutmeldingen)
Onze richtlijnen rondom foutmeldingen in formulieren.

[Writing - GOV.UK](https://design-system.service.gov.uk/components/error-message/#be-clear-and-concise)
Richtlijnen van GOV.UK over het schrijven van foutmeldingen.

## Onderzoek

[Improve Validation Errors with Adaptive Messages - Baymard Institute](https://baymard.com/blog/adaptive-validation-error-messages)
Artikel waarin wordt geadviseerd om de error messages 'op maat' te schrijven.

Eerder deden we een poging om tot een patroon voor foutmeldingen te komen. Dit was voor het prototype van het formulier dat werd gebruikt voor WMEBV. [Bekijk de poging in Figma](https://www.figma.com/file/iIr1gkAR3oZ0UFWKIni8Nv/VNG---WMEBV---Templates?type=design&node-id=854-20841&mode=design&t=TqOj9POiZQfOFkIw-4) inclusief afwegingen als comments. Deze zijn ingezet bij een onderzoek waarbij we spraken met mensen met een beperking of die minder taalvaardig waren. De resultaten van het onderzoek staan op [gebruikersonderzoeken.nl](https://gebruikersonderzoeken.nl/docs/onderzoek-bekijken/formulieren/vng-online-formulieren). Zoek daar bijvoorbeeld op 'foutmeldingen'.

## Handige linkjes

[WCAG-succescriterium 3.3.1 Foutidentificatie - NL Design System](https://www.nldesignsystem.nl/wcag/3.3.1/)
WCAG richtlijn zoals deze beschikbaar is op nldesignsystem.nl

[Toegankelijke foutmeldingen in formulieren - NL Design System](https://nldesignsystem.nl/blog/toegankelijke-foutmeldingen-formulieren/)
Artikel door Rian Rietveld waarin ook kort wordt ingegaan op het schrijven van foutmeldingen.

[Spreadsheet met foutmeldingen - Logius](https://docs.google.com/spreadsheets/d/1WGgvcCP0qADMN9PU8z6InRzZpVXZfGBg/edit?usp=sharing&ouid=109927640818009397098&rtpof=true&sd=true)
Spreadsheet opgezet door Aline Nap voor foutmeldingen bij Logius.

[Spreadsheet Schrijfwijze van foutmeldingen](https://docs.google.com/spreadsheets/d/1h5ITOrrd9MeN9MtxwPNN9QbHubvH_Ep3_bikB-QyCZ4/edit?usp=sharing)
Spreadsheet opgezet door Rozerin als poging tot een opzet voor de foutmeldingen die eerder in deze Discussion zijn genoemd. Het bevat een verzameling van de informatie die hieronder al is besproken, aangevuld met concrete voorbeelden van foutmeldingen.

**Comment by jeffreylauwers** [comment]
# Eerste samenkomst

Op 19 december hebben we met de gemeenten Rotterdam, Groningen en Utrecht gezeten om gezamenlijk na te denken over het verbeteren van foutmeldingen. We hebben ideeën en best practices uitgewisseld om foutmeldingen duidelijker en gebruiksvriendelijker te maken.

https://miro.com/app/board/uXjVK_Pu_OY=/?moveToWidget=3458764610269507012&cot=14

Kunnen we toewerken naar een patroon dat herbruikbaar en uitbreidbaar is?

**Comment by JeroenduC** [comment]
Liked-subscribed.

Leuk om te kijken of we een set aan veelgebruikte velden kunnen linken aan foutmeldingen + regEx, zodat we daar een standaard in kunnen maken.

**Comment by jeffreylauwers** [comment]
# Jeff's kladblok

Ik gebruik deze comment om mijn eigen gedachtes rond foutmeldingen te ordenen en aan te scherpen na feedback.

## Verplicht veld niet ingevuld?
'Vul een {onderwerp} in.'

Bijvoorbeeld:
- 'Vul een e-mailadres in.'
- 'Vul een KVK-nummer in.'
- 'Vul een voornaam in.'

## Ingevuld maar ongeldig?
Minimale foutmelding als het niet mogelijk is om een reden of goed voorbeeld aan te dragen.

'Ingevulde {onderwerp} is niet toegestaan.'

Bijvoorbeeld:
- 'Ingevulde e-mailadres is niet toegestaan.'
- 'Ingevulde KVK-nummer is niet toegestaan.'
- 'Ingevulde voornaam is niet toegestaan.'

## Ingevuld maar ongeldig? + Reden
Als het duidelijk is wat er fout is, kunnen we de reden van de foutmelding als opdracht schrijven.

'Ingevulde {onderwerp} is niet toegestaan. {reden}.'

Bijvoorbeeld:
- 'Ingevulde e-mailadres is niet toegestaan. Er mist een @.'
- 'Ingevulde KVK-nummer is niet toegestaan. Gebruik alleen cijfers'
- 'Ingevulde voornaam is niet toegestaan. Gebruik alleen letters.'
- 'Ingevulde datum is niet toegestaan. De datum moet in de toekomst liggen.'

## Ingevuld maar ongeldig? + Goed voorbeeld
Als de data op een bepaalde manier ingevuld moet worden, en er geen voorbeeld als Form Field Description beschikbaar is, kunnen we een goed voorbeeld tonen.

'Ingevulde {onderwerp} is niet toegestaan. Bijvoorbeeld {goed voorbeeld}.'

Bijvoorbeeld:
- 'Ingevulde e-mailadres is niet toegestaan. Vul een e-mailadres in, zoals bijvoorbeeld naam@emailadres.nl'
- 'Ingevulde postcode is niet toegestaan. Vul een nederlandse postcode in, zoals bijvoorbeeld 1234AB'

## Patroon per type?
Is het een idee om patronen te maken op basis van type foutmelding?

- Leeg --> Die hebben we al.
- Te lang --> ‘{onderwerp} moet {aantal} tekens of minder zijn.
- Te kort --> ‘{onderwerp} moet {aantal} tekens of meer zijn.
- Gebruik van karakters die niet zijn toegestaan  --> '{onderwerp} mag geen {karakter} hebben.
- Verkeerde format

## Patroon per onderwerp?
Is het een idee om patronen te maken op basis uitgevraagde data?

- Naam
- E-mailadres
- KVK-nummer
- etc.

## Patroon per input type?
Is het een idee om patronen te maken op basis van input type?

- Text Input --> Vul in
- Checkbox Group --> 'Kies...'
- Radio Group --> 'Kies...'
- Select --> 'Kies...'
- File Input --> 'Het gekozen bestand is te groot.' of 'Het gekozen bestandstype is niet toegestaan.'

['Selecteer' is geen B1 woord](https://ishetb1.nl/search?word=selecteer)

## Notities

- Ik probeer bij bovenstaande patronen continu te onthouden dat ze ook als opsomming / samenvatting van alle fouten in een [Invalid Form Alert](https://github.com/orgs/nl-design-system/discussions/278) kunnen staan.
- Expres niet 'uw' omdat het niet altijd de gegevens zijn van de persoon die de data invult.
- Over het woord 'Geldig':
  - ['Geldig is B1](https://ishetb1.nl/search?word=Geldig).
  - 'Geldig' zorgt ervoor dat je soms ook 'Geldige' krijgt. Bijvoorbeeld 'Vul een geldig e-mailadres in' en 'Vul een geldige voornaam in'.
  - 'Geldig' is misschien niet het juiste woord? Door een te strenge afstemming van validatie kan iets dat wel degelijk geldig is zoals een naam die niet vaak voorkomt, als ongeldig worden gezien. En daarmee leg je de schuld bij de gebruiker.
  - ['Correct' is geen B1](https://ishetb1.nl/search?word=correct).
- Misschien iets met 'niet toegestaan'? 
- Uit een onderzoek met mensen met een beperking en laaggeletterden werden [foutmeldingen als duidelijk ervaren](https://gebruikersonderzoeken.nl/docs/onderzoek-bekijken/formulieren/vng-online-formulieren#:~:text=formulier%20werden%20als-,duidelijk%20ervaren,-(Figuur%206).%20Dit). Hierbij werd de zin 'De ingevulde {onderwerp} is niet toegestaan. {aanvulling}' gebruikt.
- 'De ingevulde...' gaat echter niet op bij e-mailadres of KVK-nummer. Hier zou het 'Het ingevulde...' moeten zijn. Is dit per {onderwerp} in te stellen? Voor nu laat ik 'De/Het' in zijn geheel weg.
- Wanneer het 'goede voorbeeld' al als Form Field Description staat aangegeven, dan niet in de foutmelding herhalen.
- Een punt aan het eind van de zin  is verstandig. Zo kunnen schermlezers even een pauze laten vallen. Zou dat bij het aanreiken van goede voorbeelden verkeerd kunnen uitpakken? 'Bijvoorbeeld naam@emailadres.nl.'

**Comment by Rozerinay** [comment]
Hi allemaal! Ik heb een Google Spreadsheet opgesteld als poging tot een opzet voor de foutmeldingen die eerder in deze Discussion zijn genoemd. Het bevat een verzameling van de informatie die hierboven al besproken is, aangevuld met concrete voorbeelden van foutmeldingen.

Dit document is bedoeld als een eerste versie, en er staan ook nog foutmeldingen in waarbij keuzes gemaakt moeten worden. Niets is definitief, en ik hoor graag jullie feedback of suggesties om dit verder te verbeteren.

Ik zit eraan te denken om in verschillende sheets binnen de spreadsheet verschillende patronen uit te werken, zodat alles overzichtelijk blijft.

[Spreadsheet Schrijfwijze van foutmeldingen](https://docs.google.com/spreadsheets/d/1h5ITOrrd9MeN9MtxwPNN9QbHubvH_Ep3_bikB-QyCZ4/edit?usp=sharing)

**Reply by Rozerinay** [reply]
Wil je bijdragen? Laat het me weten, dan voeg ik je toe aan de spreadsheet zodat je bewerkingen kunt doen 😁

**Comment by rroose-rvo** [comment]
Onze hoofdredactie heeft ook een stukje opgenomen over [foutmeldingen in onze schrijfwijzer](https://nl-design-system.github.io/rvo/docs/schrijfwijzer/foutmeldingsteksten/).

**Comment by JeroenduC** [comment]
E-mail adres helemaal uitgeplozen:

1. Missing @ symbol or multiple @'s
2. Invalid characters in username/domain
3. Empty parts (username/domain/TLD)
4. Invalid starting/ending characters
5. Consecutive dots
6. Length limits
7. TLD format

Met een stukje AI-JavaScript er bij. Use with care.

En dan zou je dus per categorie de juiste schrijfwijze vinden. Wel veel gedoe, maar omdat e-mail adres zo vaak voor komt is dit misschien wel de moeite waard om duidelijk te maken waar de fout zit.

```
class EmailValidator {
  constructor(email) {
    this.email = email.trim();
    this.errors = [];
  }

  validate() {
    this.checkEmpty();
    this.checkAtSymbol();
    this.checkLocalPart();
    this.checkDomain();
    this.checkTLD();
    this.checkSpecialCharacters();
    this.checkLength();
    return {
      isValid: this.errors.length === 0,
      errors: this.errors
    };
  }

  checkEmpty() {
    if (!this.email) {
      this.errors.push('Email address cannot be empty');
    }
  }

  checkAtSymbol() {
    const atCount = (this.email.match(/@/g) || []).length;
    if (atCount === 0) {
      this.errors.push('Email must contain an @ symbol');
    } else if (atCount > 1) {
      this.errors.push('Email cannot contain multiple @ symbols');
    }
  }

  checkLocalPart() {
    const localPart = this.email.split('@')[0];
    if (!localPart) {
      this.errors.push('Username part before @ cannot be empty');
    } else if (localPart.length > 64) {
      this.errors.push('Username part cannot exceed 64 characters');
    }

    if (/^[.-]/.test(localPart)) {
      this.errors.push('Username cannot start with a dot or hyphen');
    }
    if (/[.-]$/.test(localPart)) {
      this.errors.push('Username cannot end with a dot or hyphen');
    }
  }

  checkDomain() {
    const parts = this.email.split('@');
    if (parts.length === 2) {
      const domain = parts[1];
      if (!domain) {
        this.errors.push('Domain part after @ cannot be empty');
      } else if (!/^[a-zA-Z0-9.-]+$/.test(domain)) {
        this.errors.push('Domain can only contain letters, numbers, dots, and hyphens');
      }
      if (/^[.-]/.test(domain)) {
        this.errors.push('Domain cannot start with a dot or hyphen');
      }
      if (/[.-]$/.test(domain)) {
        this.errors.push('Domain cannot end with a dot or hyphen');
      }
    }
  }

  checkTLD() {
    const parts = this.email.split('.');
    const tld = parts[parts.length - 1];
    if (!tld) {
      this.errors.push('Email must have a top-level domain');
    } else if (tld.length < 2) {
      this.errors.push('Top-level domain must be at least 2 characters');
    } else if (!/^[a-zA-Z]+$/.test(tld)) {
      this.errors.push('Top-level domain can only contain letters');
    }
  }

  checkSpecialCharacters() {
    const localPart = this.email.split('@')[0];
    if (!/^[a-zA-Z0-9._-]+$/.test(localPart)) {
      this.errors.push('Username can only contain letters, numbers, dots, underscores, and hyphens');
    }
    if (/[.]{2,}/.test(this.email)) {
      this.errors.push('Email cannot contain consecutive dots');
    }
  }

  checkLength() {
    if (this.email.length > 254) {
      this.errors.push('Email address cannot exceed 254 characters');
    }
  }
}

// Usage example:
const validator = new EmailValidator('test..user@domain..com');
const result = validator.validate();
console.log(result.isValid); // false
console.log(result.errors);  // Array of specific error messages
```

**Comment by jeffreylauwers** [comment]
Tijdens de Design Open Dag van september 2026 kwam de schrijfwijze van een foutmelding bij 'verplichte velden' ook ter sprake. Daar werd opgemerkt dat iemand ook wel eens de volgende melding had gekregen.

Vul {onderwerp} in om {CTA}.

Waar {onderwerp} dan dan tekst van het label was, zoals bijvoorbeeld voornaam of huisnummer. En op de plek van {CTA} een afgeleide tekst stond van de tekst op de submit button, zoals bijvoorbeeld 'Doorgaan' of 'Volgende stap'.

Dit resulteerd dan in foutmeldingen als:

- Vul voornaam in om door te gaan.
- Vul huisnummer in om door te gaan.
